File: src/main/java/komu/blunt/objects/TypeConstructorValue.java
package komu.blunt.objects;

import java.util.Arrays;

import static com.google.common.base.Preconditions.checkNotNull;

public final class TypeConstructorValue {
    
    private final String name;
    public final Object[] items;

    public TypeConstructorValue(String name, Object[] items) {
        this.name = checkNotNull(name);
        this.items = checkNotNull(items);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        
        if (name.equals("()")) {
            // TODO: unit is not really handled here for now
            return "()";
        } else if (isTuple()) {
            sb.append('(');
            for (int i = 0; i < items.length; i++) {
                if (i != 0) sb.append(", ");
    
                sb.append(items[i]);
            }
            sb.append(')');
        } else {
            sb.append('(').append(name);
            
            for (Object param : items)
                sb.append(' ').append(param);

            sb.append(')');
        }
        
        return sb.toString();
    }

    public boolean isTuple() {
        return name.startsWith("(,");
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        
        if (obj instanceof TypeConstructorValue) {
            TypeConstructorValue rhs = (TypeConstructorValue) obj;
            
            return name.equals(rhs.name)
                && Arrays.equals(items, rhs.items);
        }
        
        return false;
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(items);
    }
}

File: src/main/kotlin/komu/blunt/objects/TypeConstructorValue.kt
package komu.blunt.objects

import java.util.*

class TypeConstructorValue(val index: Int, val name: String, val items: Array<Any?> = TypeConstructorValue.EMPTY_ARRAY) : Comparable<TypeConstructorValue> {

    init {
        require(index >= 0) { "invalid index: $index" }
    }

    companion object {
        private val EMPTY_ARRAY = Array<Any?>(0) { x -> null }
    }

    fun isTuple() = name.startsWith("(,")

    override fun toString(): String =
        when {
            items.size == 0 -> name
            isTuple()       -> toStringAsTuple()
            name == ":"     -> toStringAsList()
            else            -> toStringDefault()
        }

    private fun toStringAsList(): String {
        val sb = StringBuilder()

        sb.append('[')

        sb.append(items[0])

        var value: TypeConstructorValue = items[1] as TypeConstructorValue
        while (value.name == ":") {
            sb.append(", ").append(value.items[0])
            value = value.items[1] as TypeConstructorValue
        }

        sb.append(']')

        return sb.toString()
    }

    private fun toStringDefault(): String {
        val sb = StringBuilder()

        sb.append('(').append(name)

        for (param in items)
            sb.append(' ').append(param)

        sb.append(')')
        return sb.toString()
    }

    private fun toStringAsTuple() =
        items.joinToString(", ", "(", ")")

    override fun equals(o: Any?) =
        o is TypeConstructorValue && index == o.index && Arrays.equals(items, o.items)

    override fun hashCode() =
        Arrays.hashCode(items)

    override fun compareTo(other: TypeConstructorValue): Int {
//        if (index != o.index)
//            return index - o.index;
//
//        for (i in items.indices) {
//            val lhs = items[i] as Comparable<*>
//            val rhs = o.items[i] as Comparable<*>
//            val c = lhs.compareTo(rhs);
//            if (c != 0)
//                return c;
//        }
//        return 0;
        throw UnsupportedOperationException("comparison is not supported")
    }
}
